packages/core/src/browser/markdown-renderer.ts

```diff
@@ -0,0 +1,74 @@
+/********************************************************************************
+ * Copyright (C) 2021 TypeFox and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as DOMPurify from 'dompurify';
+import * as markdownit from 'markdown-it';
+
+export class MarkdownRenderer {
+
+    protected engine: markdownit;
+    protected domParser = new DOMParser();
+    protected callbacks = new Map<string, ((element: Element) => Element | void)[]>();
+
+    constructor(engine?: markdownit) {
+        this.engine = engine ?? markdownit();
+    }
+
+    /**
+     * Adds a modification callback that is applied to every element with the specified tag after rendering to HTML.
+     *
+     * @param tag The tag that this modification applies to
+     * @param callback The modification to apply on every selected rendered element. Can either modify the element in place or return a new element.
+     */
+    modify<K extends keyof HTMLElementTagNameMap>(tag: K, callback: (element: HTMLElementTagNameMap[K]) => Element | void): MarkdownRenderer {
+        if (this.callbacks.has(tag)) {
+            this.callbacks.get(tag)!.push(callback);
+        } else {
+            this.callbacks.set(tag, [callback]);
+        }
+        return this;
+    }
+
+    render(markdown: string): HTMLElement {
+        return this.renderInternal(this.engine.render(markdown));
+    }
+
+    renderInline(markdown: string): HTMLElement {
+        return this.renderInternal(this.engine.renderInline(markdown));
+    }
+
+    protected renderInternal(renderedHtml: string): HTMLElement {
+        const div = this.sanitizeHtml(renderedHtml);
+        for (const [tag, calls] of this.callbacks) {
+            for (const callback of calls) {
+                const elements: Element[] = Array.from(div.getElementsByTagName(tag));
+                for (const element of elements) {
+                    const result = callback(element);
+                    if (result) {
+                        div.replaceChild(result, element);
```

Comment:
It's still pretty easy to get this to fail. `.getElementsByTagName` retrieves all descendants, but (in Chrome, at least) `replaceChild` only works if the element to be replaced is an immediate child of the element on which `replaceChild` is called, so if the original rendering or a previous callback creates a nested structure that includes the tag of interest, the next callback targeting that tag will fail. Since `markdown-it` already supports plugins, I wouldn't make this hard on yourself trying to invent a new plugin system for it: just apply the modifications you want to apply to the HTML.

Reply:
I would like to use the plugin system of `markdown-it`, but it's just not powerful enough. The main issue is that it always renders to text, eliminating every listener I place on the html elements in the process. The main purpose of the `MarkdownRenderer` class is to operate directly on the html elements. (I've tried implementing it using purely `markdown-it` at the start, but couldn't get around this limitation)

Anyway, changed this to only select direct children of the `div`.

Reply:
Thanks for adding the tests. I don't see an obvious way to break the current implementation (though a triply nested for-loop is a little scary...) 👍. I'm still not sure whether it's worth adding that machinery when the only current use case is a fairly simple one (running one callback on each of two different tags). Adding it puts us on the hook for its correctness in the general case when that level of complexity isn't (yet) needed. But it is working, and the tests suggest it should handle the general case, so I'll leave it up to you.